lib/Monads/trace/Trace_Reader_Option.thy

```diff
@@ -0,0 +1,217 @@
+
+lemma gets_the_if_distrib:
+  "gets_the (if P then f else g) = (if P then gets_the f else gets_the g)"
+  by simp
+
+lemma gets_the_oapply_comp:
+  "gets_the (oapply x \<circ> f) = gets_map f x"
+  by (fastforce simp: gets_map_def gets_the_def o_def gets_def)
+
+lemma gets_the_Some:
+  "gets_the (\<lambda>_. Some x) = return x"
+  by (simp add: gets_the_def assert_opt_Some)
+
+lemma gets_the_oapply2_comp:
+  "gets_the (oapply2 y x \<circ> f) = gets_map (swp f y) x"
+  by (clarsimp simp: gets_map_def gets_the_def o_def gets_def)
+
+lemma gets_obind_bind_eq:
+  "(gets (f |>> (\<lambda>x. g x))) =
+   (gets f >>= (\<lambda>x. case x of None \<Rightarrow> return None | Some y \<Rightarrow> gets (g y)))"
+  by (auto simp: simpler_gets_def bind_def obind_def return_def split: option.splits)
+
+lemma mres_assert_opt:
+  "mres (assert_opt opt s) = (if opt = None then {} else {(the opt,s)})"
+  by (clarsimp simp: assert_opt_def fail_def return_def mres_def vimage_def split: option.split)
+
+
+lemmas omonad_simps [simp] =
+  gets_the_opt_map assert_opt_Some gets_the_obind
+  gets_the_return gets_the_fail gets_the_returnOk
+  gets_the_throwError gets_the_assert gets_the_Some
+  gets_the_oapply_comp
+
+lemmas in_omonad = bind_eq_Some_conv in_obind_eq in_opt_map_eq in_opt_pred Let_def
+
+
+section "Relation between option monad loops and trace monad loops."
+
+(* Option monad whileLoop formalisation thanks to Lars Noschinski <noschinl@in.tum.de>. *)
+
+lemma gets_the_conv:
+  "(gets_the B s) = (case B s of Some r' \<Rightarrow> ({([], Result (r', s))}) | _ \<Rightarrow> {([], Failed)})"
+  by (auto simp: gets_the_def gets_def get_def bind_def return_def fail_def assert_opt_def split: option.splits)
+
+lemma gets_the_loop_terminates:
+  "whileLoop_terminates C (\<lambda>a. gets_the (B a)) r s
+    \<longleftrightarrow> (\<exists>rs'. (Some r, rs') \<in> option_while' (\<lambda>a. C a s) (\<lambda>a. B a s))" (is "?L \<longleftrightarrow> ?R")
+proof
+  assume ?L then show ?R
+  proof (induct rule: whileLoop_terminates.induct[case_names 1 2])
+    case (2 r s) then show ?case
+      by (cases "B r s") (auto simp: gets_the_conv intro: option_while'.intros)
+  qed (auto intro: option_while'.intros)
+next
+  assume ?R then show ?L
+  proof (elim exE)
+    fix rs' assume "(Some r, rs') \<in> option_while' (\<lambda>a. C a s) (\<lambda>a. B a s)"
+    then have "whileLoop_terminates C (\<lambda>a. gets_the (B a)) (the (Some r)) s"
+      by induct (auto intro: whileLoop_terminates.intros simp: gets_the_conv)
+    then show ?thesis by simp
+  qed
+qed
+
+lemma The_eqD:
+  "\<lbrakk>The P = x; \<exists>!x. P x\<rbrakk> \<Longrightarrow> P x"
+  by (metis theI)
+
+lemma gets_the_whileLoop:
+  fixes C :: "'a \<Rightarrow> 's \<Rightarrow> bool"
+  assumes terminates: "\<forall>s. whileLoop_terminates C (\<lambda>a. gets_the (B a)) r s"
+  shows "whileLoop C (\<lambda>a. gets_the (B a)) r = gets_the (owhile C B r)"
```

Comment:
Non-terminating loops are a bit more complicated for the trace monad compared to the nondet monad, since I don't think that we can 'just' fail. I'm pretty confident that that means we need this extra assumption here, but it's possible that with more thought there could be a change made to the definition of `whileLoop` instead.

Reply:
I think this is right. Non-terminating loops make sense in the trace monad setting (producing an unbounded trace set), so mapping them to failure is no longer the right thing to do, which means the non-termination case no longer matches up with the reader monad.

There is probably a tweak possible that we can do later: since we're comparing with the reader monad, we know there is no state change and (I think) no trace produced. In that specific case, failure might actually make sense for the while loop, but I'm not clear whether that would come out of the current definitions or not.

Reply:
Ok, definitely something to come back to later.